Beta README updates (#12)
americanexpress/fetchye: README.md

## ✨ Features

* ES6 Fetch powered by React Hooks
* Pure React or Redux Shared Cache modes available
* Headless per-Hook Cache Mode available
* SSR-friendly

## ⬇️ Install & Setup

**Contents**
* [Quick Install](#quick-install)
* [FetchyeProvider Install](#fetchyeprovider-install)
* [FetchyeReduxProvider Install](#fetchyereduxprovider-install)
* [One App Install](#one-app-install)

### Quick Install

>💡 Makes use of Headless per-Hook Cache Mode

**Pros**
- Painless and Quick

**Cons**
- No shared caching
- No de-duplication of API calls

Just install and do `useFetchye`. Thats it!

```sh
npm i -S fetchye
```

```js
// ...
import { useFetchye } from 'fetchye';

const MyComponent = () => {
  const { isLoading, data } = useFetchye('http://example.com/api/profile');
  return (
    <>
      {!isLoading && (
        <p>{data?.body.name}</p>
      )}
    </>
  );
};
```

🏖️[Try this out on Codesandbox](https://codesandbox.io/s/americanexpressfetchye-quick-install-w4ne2)

### `FetchyeProvider` Install

>💡 When you want a central cache but no extra dependencies

**Pros**
- Easy
- Shared Cache
- De-duplication of API calls

**Cons**
- No Redux Dev Tools for debugging and cache inspection

> 

Author: Fixed in 3cacd66

```js
// ...
import { useFetchye } from 'fetchye';

const MyComponent = () => {
  const { isLoading, data } = useFetchye('http://example.com/api/profile');
  return (
    <>
      {!isLoading && (
        <p>{data?.body.name}</p>
      )}
    </>
  );
};
```

🏖️[Try this out on Codesandbox](https://codesandbox.io/s/americanexpressfetchye-fetchye-provider-install-y7d8j)

### `FetchyeReduxProvider` Install

>💡 When you want a central cache integrated with a Redux based project

**Pros**
- Easy if you know Redux
- Shared Cache
- De-duplication of API calls
- Redux Dev Tools for debugging and cache inspection

**Cons**
- More steps and dependencies

Add `fetchye` and its needed optional dependencies:

```
npm i -S fetchye redux react-redux
```

Add the `<FetchyeReduxProvider />` component under the Redux `<Provider />`:

```jsx
import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { FetchyeReduxProvider } from 'fetchye/lib/FetchyeReduxProvider';
import { SimpleCache } from 'fetchye/lib/cache';

const fetchyeCache = SimpleCache({
  // Need to tell Fetchye where the cache reducer will be located

>   cacheSelector: (state) => state,

Author: The following will break the example
```
const fetchyeCache = SimpleCache({
  // Need to tell Fetchye where the cache reducer will be located
  cacheSelector: (state) => state. fetchCache,
});
const store = createStore(fetchyeCache.reducer);
```

Not going to change this. Addressed a One App specific install in the following section.

## ✨ Features

* ES6 Fetch powered by React Hooks
* Pure React or Redux Shared Cache modes available
* Headless per-Hook Cache Mode available
* SSR-friendly

> ## ⬇️ Install

Reviewer: Should this be `🤹‍ Usage` instead? This section goes over what to install and how to use it

Author: No. This is how to install these. Usage is farther down in another section.

Reviewer: I will settle for `Install & Setup` or `Install and Setup` whichever makes a nicer link

Author: https://github.com/americanexpress/fetchye/blob/docs/readme/README.md#%EF%B8%8F-install--setup

Author: Fixed in https://github.com/americanexpress/fetchye/pull/12/commits/7c5e59763561b539fa6c985bcad2e78ce539435e

};
```

> ### Sequential API Execution

Reviewer 2: nice :-D

};
```

### Custom Fetcher

Custom fetchers allow for creating reusable data fetching logic for specific

> APIs or custom needs. They also allow for accepting a centrally provided
> `fetchClient` but wrapping that client on a per `useFetchye` request basis.

Reviewer 3: ```suggestion
APIs or custom needs. They allow for a centrally provided
`fetchClient` which wraps that client on a per `useFetchye` request basis.
```

Author: Fixed in 3cacd66

    // dispatches events into the server side Redux store
    await fetchye('http://example.com/api/books/');
  },
};

export default BookList;
```

#### Next.JS SSR

```jsx
import SimpleCache from 'fetchye/lib/cache/SimpleCache';
import { useFetchye, makeServerFetchye } from 'fetchye';

const cache = SimpleCache();

// Codesandbox takes a second to get Next.JS started...
export default function IndexPage({ initialBookList }) {
  const { isLoading, data } = useFetchye('http://example.com/api/books/', {
    initialData: initialBookList,
  });

  if (isLoading) {
    return (<p>Loading...</p>);
  }

  return (
    <>
      {/* Render data */}
    </>
  );
}

const fetchye = makeServerFetchye({
  cache,
  fetchClient: fetch,
});

export async function getServerSideProps() {
  try {
    // returns { data, error } payload for feeding initialData on client side
    const res = await fetchye('http://example.com/api/books/');
    return {
      props: {
        initialBookList: res,
      },
    };
  } catch (error) {
    console.error(error.message);
    return {};
  }
}
```

🏖️[Try this out on Codesandbox](https://codesandbox.io/s/americanexpressfetchye-fetchye-ssr-0ktx9)

> ## 🎛️ API

Reviewer 2: What do you think about having something similar to this https://github.com/nellyk/holocron/tree/main/packages/holocron/docs/api

Reviewer 2: especially with the table of contents

Author: Yes, it definitely could use a table of contents.

Author: Fixed in https://github.com/americanexpress/fetchye/commit/3abb96435114948636c3f657a0364b265568ff91

### `useFetchye`

> 

Reviewer 2: I know we've mentioned about when to use this at the top, but it would be good to get a description here too

Author: I could add a short description. Most of the information for usage will come from previous sections however.

Author: Fixed in https://github.com/americanexpress/fetchye/commit/3abb96435114948636c3f657a0364b265568ff91
